gnucash/gnome/gnc-budget-view.c

```diff
@@ -1190,33 +1190,73 @@ totals_col_source (GtkTreeViewColumn *col, GtkCellRenderer *cell,
     {
         account  = gnc_account_nth_child (priv->rootAcct, i);
         currency = gnc_account_get_currency_or_parent (account);
-        neg = FALSE;
 
-        switch (xaccAccountGetType (account))
-        {
-        case ACCT_TYPE_ASSET:
-        case ACCT_TYPE_LIABILITY:
-        case ACCT_TYPE_EQUITY:
-            if (row_type != TOTALS_TYPE_ASSET_LIAB_EQ &&
-                row_type != TOTALS_TYPE_REMAINDER)
-                continue;
-            neg = (row_type != TOTALS_TYPE_ASSET_LIAB_EQ);
-            break;
-        case ACCT_TYPE_INCOME:
-            if (row_type != TOTALS_TYPE_INCOME &&
-                row_type != TOTALS_TYPE_REMAINDER)
-                continue;
-            neg = (row_type == TOTALS_TYPE_ASSET_LIAB_EQ);
-            break;
-        case ACCT_TYPE_EXPENSE:
-            if ((row_type != TOTALS_TYPE_EXPENSES) &&
-                (row_type != TOTALS_TYPE_REMAINDER))
-                continue;
-            neg = (row_type == TOTALS_TYPE_REMAINDER ||
-                   row_type == TOTALS_TYPE_ASSET_LIAB_EQ);
-            break;
-        default:
-            continue;
+        if (gnc_using_unreversed_budgets(gnc_account_get_book(account)))
+        {  /* using book with unreversed-budgets feature. This will be
+              the default in 4.x after budget scrubbing*/
+            neg = gnc_reverse_balance (account);
+            switch (xaccAccountGetType (account))
+            {
+            case ACCT_TYPE_ASSET:
+                if (row_type != TOTALS_TYPE_ASSET_LIAB_EQ &&
+                    row_type != TOTALS_TYPE_REMAINDER)
+                    continue;
```

Comment:
While I can't verify if all the negations are correct, this switch statement struck me as very difficult to reason about: almost each case starts with a condition consisting of two negative tests 'and'-ed together. One of these is about a total type not directly related to the account type.
I have rewritten the switch statements in [the clam-budget branch](https://github.com/gjanssens/gnucash/commit/4f624087e1e0c521aca46254633b3ae1ad7d5d78) of my personal repo. The rewrite switches on row_type rather than account type. The logic that follows from that arrangement looks easier to parse by humans. What do you think ?

Reply:
Agree. #503 had twisted some easy conditionals into this switch mess. I'll pull from your branch soon.

If you have time testing the negations, the attached budget-featured.zip is the test book.